books/kestrel/c/syntax/types.lisp

```diff
@@ -850,172 +906,471 @@
+
+  (define type-params-compatiblep ((x type-paramsp)
+                                   (y type-paramsp)
+                                   (ienv ienvp))
+    :returns (yes/no booleanp)
+    :short "Check that parameter portions of two function @(see type)s are
+            compatible [C17:6.2.7]."
+    :long
+    (xdoc::topstring
+     (xdoc::p
+       "This is the second part of the compatibility check on function types.
+        In addition to the earlier check
+        that the return types must be compatible,
+        the following conditions must also hold."
+       (xdoc::ol
+        (xdoc::li
+         "When both function types correspond to function prototypes,
+          both must have the same number of parameters
+          and each parameter must have compatible type.
+          Furthermore, both must agree on the ellipsis terminator.")
+        (xdoc::li
+         "When one type corresponds to a function prototype
+          and the other is part of an ``old-style'' function definition,
+          each type in the parameter type list of the prototype
+          must be compatible with the corresponding of the old-style function
+          after the latter has gone through default argument promotion.
+          Furthermore, the prototype must not feature an ellipsis terminator.")
+        (xdoc::li
+         "When one type corresponds to a function prototype
+          and the other has an empty identifier list
+          and is not derived from a function definition,
+          the type of each parameter in the function prototype
+          must be compatible with itself after default argument promotion.
+          Furthermore, the prototype must not feature
+          an ellipsis terminator."))
+       "In the above mention of parameter types,
+        we mean the type after adjustment [C17:6.7.6.3/7-8].
+        This requires no special consideration here,
+        since we always represent function types post-adjustment
+        (see @(see type))."))
+    (type-params-case
+      x
+      :prototype
+      (type-params-case
+        y
+        :prototype (and (type-list-compatiblep x.params y.params ienv)
+                        (equal x.ellipsis y.ellipsis))
+        :old-style (and (not x.ellipsis)
+                        (type-list-compatiblep
+                          x.params
+                          (type-list-da-promote y.params ienv)
+                          ienv))
+        :unspecified (and (not x.ellipsis)
+                          (type-list-compatiblep
+                            x.params
+                            (type-list-da-promote x.params ienv)
+                            ienv)))
+      :old-style
+      (type-params-case
+        y
+        :prototype (and (not y.ellipsis)
+                        (type-list-compatiblep
+                          (type-list-da-promote x.params ienv)
+                          y.params
+                          ienv))
```

Comment:
It looks like there is no requirement between `:old-style` and `:old-style`. This looks consistent with [C17], which simply says nothing about that situation, as opposed to giving requirements for the other situations. It may be worth mentioning this explicitly in the documentation of this function (either as another `xdoc::li` or as text after the `xdoc::ol`), because it took me a bit to realize this, and I expect others may experience the same.